### PR TITLE
Adding resolve_offset/6 to brod_utils, as a way to pass a custom time…

### PR DESCRIPTION
**Pull request by escobera**
…out to kpro:connect_partition_leader

### CODE REVIEW

**Comment by zmstone** [comment]
Please also export the API from brod.erl

File: src/brod_utils.erl [hunk 1]
```diff
@@ -141,8 +143,17 @@ get_metadata(Hosts, Topics, ConnCfg) ->
                      offset_time(), conn_config()) ->
         {ok, offset()} | {error, any()}.
 resolve_offset(Hosts, Topic, Partition, Time, ConnCfg) ->
+  Opts = #{timeout => proplists:get_value(
+                                  connect_timeout, ConnCfg, ?DEFAULT_TIMEOUT)},
```
Comment (zmstone) [inline src/brod_utils.erl]: elvis is not happy about this line being too long (see travis build),
maybe add a variable for the fallback timeout value.

**Comment by escobera** [comment]
> Please also export the API from brod.erl

Hm, that got me thinking. Will `brod_utils:resolve_offset/6` be public? I don't really know much about erlang so maybe I'm getting this wrong, but this seems to be a private method which only serves as a way to pass connect_timeout property.

In any case I tried exposing it but couldn't get it to pass the dialyzer test. Could you give me some guidance here?

**Comment by zmstone** [comment]
Hi. 
`brod_utils` is a private (by convention) module although there is no way to stop other applications from calling its exported functions directly.
Most of the public APIs in `brod` are exported from brod.erl, the calls to other modules can be simply a delegation.
Worth pointing out that there is no runtime overhead for such delegation thanks to compiler optimisation.

**Comment by zmstone** [comment]
Should be added to the export list in bord.erl.
https://github.com/klarna/brod/blob/master/src/brod.erl#L94

**Comment by zmstone** [comment]
and add a test case to cover the new code:
e.g.
https://github.com/klarna/brod/blob/master/test/brod_consumer_SUITE.erl#L357

File: src/brod.erl [hunk 2]
```diff
@@ -850,7 +850,18 @@ resolve_offset(Hosts, Topic, Partition, Time) ->
                      offset_time(), conn_config()) ->
         {ok, offset()} | {error, any()}.
 resolve_offset(Hosts, Topic, Partition, Time, ConnCfg) ->
-  brod_utils:resolve_offset(Hosts, Topic, Partition, Time, ConnCfg).
+  Timeout =
```
Comment (zmstone) [inline src/brod.erl]: this can be forwarded to `brod_utils:resolve_offset/5`

Comment (zmstone) [reply src/brod.erl]: i.e. call `brod_utils:resolve_offset(Hosts, Topic, Partition, Time, ConnCfg).`

File: test/brod_consumer_SUITE.erl [hunk 3]
```diff
@@ -362,6 +363,23 @@ t_direct_fetch(Config) when is_list(Config) ->
...
+  Client = ?config(client),
+  Topic = ?TOPIC,
+  Partition = 0,
+  Key = make_unique_key(),
+  Value = <<>>,
+  ClientConfig = lists:merge(?config(client_config), 
+                              [{connect_timeout, timer:seconds(45)}]),
+  ok = brod:produce_sync(Client, ?TOPIC, Partition, Key, Value),
+  {ok, Offset} = brod:resolve_offset(?HOSTS, Topic, Partition,
+                                     ?OFFSET_LATEST, ClientConfig),
```
Comment (zmstone) [inline test/brod_consumer_SUITE.erl]: I mean test `brod:resolve_offset/6`

Comment (escobera) [reply test/brod_consumer_SUITE.erl]: Ok, I'll do that

File: test/brod_consumer_SUITE.erl [hunk 4]
```diff
@@ -362,6 +363,22 @@ t_direct_fetch(Config) when is_list(Config) ->
   ?assertEqual(Key, Msg#kafka_message.key),
   ok.
 
+t_direct_fetch_with_connect_timeout(Config) when is_list(Config) ->
```
Comment (zmstone) [inline test/brod_consumer_SUITE.erl]: one last thing
this test case is duplicating the above one.
Maybe rename it to `t_resolve_offset` and no need to test anything else than the call to `brod:resolve_offset/6`.
e.g.
```
t_resolve_offset(Config) when is_list(Config) ->
  Client = ?config(client),
  Topic = ?TOPIC,
  Partition = 0,
  {ok, Offset} = brod:resolve_offset(?HOSTS, Topic, Partition,
                                     ?OFFSET_LATEST, ?config(client_config),
                                     #{timeout => timer:seconds(45)}),
  ?assertTrue(is_integer(Offset)).
```

File: src/brod.erl [hunk 5]
```diff
@@ -850,7 +851,15 @@ resolve_offset(Hosts, Topic, Partition, Time) ->
                      offset_time(), conn_config()) ->
         {ok, offset()} | {error, any()}.
 resolve_offset(Hosts, Topic, Partition, Time, ConnCfg) ->
-  brod_utils:resolve_offset(Hosts, Topic, Partition, Time, ConnCfg).
+ brod_utils:resolve_offset(Hosts, Topic, Partition, Time, ConnCfg).
```
Comment (zmstone) [inline src/brod.erl]: indent two spaces.

**Comment by escobera** [comment]
@zmstone ok I think I fixed the problems you pointed. 

I'm sorry for this back and forth, it is my first contribution to an Erlang project.

Anything else you think is laking in this PR just let me know.

**Review by zmstone — APPROVED** [review]
maybe add a record in changelog.md
Looks good otherwise.
Ping @mikpe @robertoaloi for (squash) merge

**Comment by escobera** [comment]
Any news here guys? If I need to do anything else just lemme know!

Cc: @mikpe @robertoaloi @zmstone

**Review by mikpe — APPROVED** [review]

**Comment by mikpe** [comment]
Merged, added changelog entry, tagged as 3.12.0.

**Pull request merged**